### PR TITLE
JP-3728: Deprecate `Step.__call__`

### LINKED ISSUE #8748: Print deprecation warning when neither .call nor .run is specified
_Issue [JP-3728](https://jira.stsci.edu/browse/JP-3728) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

Per discussion at <https://jira.stsci.edu/browse/JP-3490> and the Sep 4 JP meeting, we would like to add a warning notice to the pipeline log when users are directly using pipeline functions without specifying either .call() or .run().  This message would be to warn users that in a future build (TBD) usage without specifying which method to use will be deprecated.

Presumably this will apply only to usage within python environments, as command-line use via strun is understood to use the .call approach that incorporates parameter reference files.

### PR DESCRIPTION
**Pull request by braingram**
This PR adds a warning to `Step.__call__` uses to target end-users that might use this expecting the step to use reference files.

Note that `Step.__call__` is used in pipelines so the warning will only appear if `Step.__call__` is used outside a pipeline (it's used on a step that doesn't have a "parent").

If `Step.__call__` is removed, some decision will need to be made about what to do for pipeline uses of `Step.__call__`.

Regression tests show only failures that exist on main:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1694/
and do not show the new warning (so there is no use of `Step.__call__` (outside of a pipeline) in jwst.

Resolves [JP-3728](https://jira.stsci.edu/browse/JP-3728)

Closes #8748

This PR addresses ...

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Review by melanieclarke — COMMENTED** [review]
As a self-contained change, this looks good to me.  Testing locally, the warning is emitted when `__call__` is used, and not when .call or .run is directly used.  I don't see it from substeps when running a pipeline.

I think more work is needed to address the differences between .call and .run, but that's outside the scope of the request in JP-3728.  I don't see anything wrong with deprecating `__call__` in the meantime.

The change log edit needs to be converted to a towncrier fragment, and I'd like to wait to see if anyone else has opinions, so I'll hold off on approval for now.

**Review by tapastro — COMMENTED** [review]
I agree with Melanie that this change should do no harm and inform users about their use of a method we don't necessarily recommend. Just need the updates to changelog -> fragment, and I'll approve and merge.

**Comment by braingram** [comment]
Thanks @melanieclarke and @tapastro for taking a look. I updated the changelog to use towncrier.

**Review by tapastro — APPROVED** [review]

**Pull request merged**
